File: frontend/src/types/habit.ts
/**
 * TypeScript interfaces for habit-related types.
 * These mirror the backend Pydantic schemas exactly.
 */

// Completion rate nested object
export interface CompletionRate {
  week: number
  month: number
  all_time: number
}

// Habit with computed statistics (GET /habits response)
export interface HabitWithStats {
  id: string
  name: string
  description: string | null
  created_at: string // ISO datetime string
  updated_at: string // ISO datetime string
  current_streak: number
  best_streak: number
  completion_rate: CompletionRate
  completed_today: boolean
}

// Create habit request
export interface HabitCreate {
  name: string
  description?: string | null
}

// Update habit request
export interface HabitUpdate {
  name?: string
  description?: string | null
}

// Habit response (POST/PUT response)
export interface HabitResponse {
  id: string
  name: string
  description: string | null
  created_at: string
  updated_at: string
}

// Completion response
export interface CompletionResponse {
  habit_id: string
  date: string // YYYY-MM-DD format
  completed: boolean
}

// Absence item in list response
export interface AbsenceItem {
  date: string // YYYY-MM-DD format
  reason: string | null
}

// Absences list response
export interface AbsencesListResponse {
  habit_id: string
  absences: AbsenceItem[]
}

// Completions list response
export interface CompletionsListResponse {
  habit_id: string
  completions: string[] // Array of YYYY-MM-DD dates
}

// Absence create request
export interface AbsenceCreate {
  date?: string // YYYY-MM-DD, defaults to today
  reason?: string | null
}

// Absence response
export interface AbsenceResponse {
  habit_id: string
  date: string
  reason: string | null
}

File: backend/app/schemas/habit.py
"""Pydantic schemas for Habit operations."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class HabitCreate(BaseModel):
    """Schema for creating a new habit."""

    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class HabitUpdate(BaseModel):
    """Schema for updating an existing habit."""

    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)


class HabitResponse(BaseModel):
    """Schema for habit response."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None
    created_at: datetime
    updated_at: datetime
